src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+    /// STEPs 1 and 2. So we proceed by first allocating preimages and images for
+    /// each slot and then, as we traverse the LEM, we add constraints to make sure
+    /// that the witness satisfies the arithmetic equations for the corresponding
+    /// slots.
+    pub fn synthesize<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &mut Store<F>,
+        global_allocator: &mut GlobalAllocator<F>,
+        slots_count: &SlotsCounter,
+        frame: &Frame<F>,
+    ) -> Result<()> {
+        let mut allocated_ptrs: HashMap<AString, AllocatedPtr<F>> = HashMap::default();
+
+        self.allocate_input(cs, store, frame, &mut allocated_ptrs)?;
+        let preallocated_outputs = LEM::allocate_output(cs, store, frame, &mut allocated_ptrs)?;
+
+        let preallocated_hash2_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash2_ptrs,
+            SlotType::Hash2,
+            slots_count.hash2,
+            store,
+        )?;
+
+        let preallocated_hash3_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash3_ptrs,
+            SlotType::Hash3,
+            slots_count.hash3,
+            store,
+        )?;
+
+        let preallocated_hash4_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash4_ptrs,
+            SlotType::Hash4,
+            slots_count.hash4,
+            store,
+        )?;
+
+        struct Globals<'a, F: LurkField, CS: ConstraintSystem<F>> {
+            cs: &'a mut CS,
+            store: &'a mut Store<F>,
+            frame: &'a Frame<F>,
+            global_allocator: &'a mut GlobalAllocator<F>,
+            allocated_ptrs: HashMap<AString, AllocatedPtr<F>>,
+            preallocated_outputs: [AllocatedPtr<F>; 3],
+            preallocated_hash2_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+            preallocated_hash3_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+            preallocated_hash4_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+        }
+
+        fn recurse<F: LurkField, CS: ConstraintSystem<F>>(
+            block: &LEMCTL,
+            concrete_path: Boolean,
+            path: Path,
+            slots_count: &mut SlotsCounter,
+            g: &mut Globals<'_, F, CS>,
+        ) -> Result<()> {
+            match block {
+                LEMCTL::Return(output_vars) => {
+                    for (i, output_var) in output_vars.iter().enumerate() {
+                        let Some(allocated_ptr) = g.allocated_ptrs.get(output_var.name()) else {
+                            bail!("{output_var} not allocated")
+                        };
+
+                        allocated_ptr
+                            .implies_ptr_equal(
+                                &mut g.cs.namespace(|| {
+                                    format!(
+                                        "{path}.implies_ptr_equal {output_var} (output_var {i})"
```

Comment:
It seems like you're including the full path in each nested namespace, which is both unwieldy and not how namespaces are meant to work. Probably instead `recurse` should be receiving a `CS` argument, and when you add to the path, you should instead be creating a nested namespace. If `path` is also being used for some *other* purpose, I haven't caught that yet.

Reply:
Done